pilot/pkg/security/authz/builder/builder.go

```diff
@@ -0,0 +1,161 @@
+	if err != nil {
+		rbacLog.Errorf("failed to create ServiceMetadata for %s: %s", serviceName, err)
+		return nil
+	}
+
+	rbacConfig := policies.RbacConfig
+	isGlobalPermissiveEnabled := rbacConfig != nil && rbacConfig.EnforcementMode == istio_rbac.EnforcementMode_PERMISSIVE
+
+	return &Builder{
+		serviceMetadata:             serviceMetadata,
+		authzPolicies:               policies,
+		isGlobalPermissiveEnabled:   isGlobalPermissiveEnabled,
+		isXDSMarshalingToAnyEnabled: isXDSMarshalingToAnyEnabled,
+	}
+}
+
+func (b *Builder) BuildHTTPFilter() *http_filter.HttpFilter {
+	rbacConfig := b.build(false /* forTCPFilter */)
+	httpConfig := http_filter.HttpFilter{
+		Name: authz_model.RBACHTTPFilterName,
+	}
+	if b.isXDSMarshalingToAnyEnabled {
+		httpConfig.ConfigType = &http_filter.HttpFilter_TypedConfig{TypedConfig: util.MessageToAny(rbacConfig)}
+	} else {
+		httpConfig.ConfigType = &http_filter.HttpFilter_Config{Config: util.MessageToStruct(rbacConfig)}
+	}
+
+	rbacLog.Debugf("built http filter config: %v", httpConfig)
+	return &httpConfig
+}
+
+func (b *Builder) BuildTCPFilter() *tcp_filter.Filter {
+	// The build function always return the config for HTTP filter, we need to extract the
+	// generated rules and set it in the config for TCP filter.
+	config := b.build(true /* forTCPFilter */)
+	rbacConfig := &tcp_config.RBAC{
+		Rules:       config.Rules,
+		ShadowRules: config.ShadowRules,
+		StatPrefix:  authz_model.RBACTCPFilterStatPrefix,
+	}
+
+	tcpConfig := tcp_filter.Filter{
+		Name: authz_model.RBACTCPFilterName,
+	}
+	if b.isXDSMarshalingToAnyEnabled {
+		tcpConfig.ConfigType = &tcp_filter.Filter_TypedConfig{TypedConfig: util.MessageToAny(rbacConfig)}
+	} else {
+		tcpConfig.ConfigType = &tcp_filter.Filter_Config{Config: util.MessageToStruct(rbacConfig)}
+	}
+
+	rbacLog.Debugf("built tcp filter config: %v", tcpConfig)
+	return &tcpConfig
+}
+
+// isServiceInList checks if a given service or namespace is found in the RbacConfig target.
+func isServiceInList(serviceHostname string, namespace string, target *istio_rbac.RbacConfig_Target) bool {
+	if target == nil {
+		return false
+	}
+	for _, ns := range target.Namespaces {
+		if namespace == ns {
+			return true
+		}
+	}
+	for _, service := range target.Services {
+		if service == serviceHostname {
+			return true
+		}
+	}
+	return false
+}
+
+func isRbacEnabled(serviceHostname string, namespace string, policies *model.AuthorizationPolicies) bool {
+	if policies == nil || policies.RbacConfig == nil {
+		return false
+	}
+
+	rbacConfig := policies.RbacConfig
+	switch rbacConfig.Mode {
+	case istio_rbac.RbacConfig_ON:
+		return true
+	case istio_rbac.RbacConfig_ON_WITH_INCLUSION:
+		return isServiceInList(serviceHostname, namespace, rbacConfig.Inclusion)
+	case istio_rbac.RbacConfig_ON_WITH_EXCLUSION:
+		return !isServiceInList(serviceHostname, namespace, rbacConfig.Exclusion)
+	default:
+		return false
+	}
+}
+
+func (b *Builder) build(forTCPFiler bool) *http_config.RBAC {
```

Comment:
Be more clear with the method names? `build` and `generate` below have a very similar meaning but they're very different.

Reply:
Good point, I renamed them to `generateFilterConfig` and `generatePolicy` respectively with some comments.